client/js/gdpr/chart.js

```diff
@@ -4,73 +4,80 @@ const RUN_SCALE_X = 30;
-  const scrollContainer = document.createElement('div');
-  scrollContainer.classList.add('chart-scroll-container');
-
-  const svg = document.createElementNS(SVG_NS, 'svg');
-  svg.setAttribute('class', 'chart');
-  svg.setAttribute(
-      'viewBox',
-      // shift down so that lines are vertically centered
-      `0 ${RUN_SCALE_Y * 0.5} ` +
-      `${dates.length * RUN_SCALE_X} ${(NUM_POSITIONS + 0.5) * RUN_SCALE_Y}`,
-  );
-  svg.append(createDefs());
-  svg.append(createGrid(dates));
-
-  scrollContainer.append(svg);
-
-  const seriesContainer = document.createElementNS(SVG_NS, 'g');
-
-  /** @type {SVGGElement[]} */
-  const seriesElements = [];
-  const historyEntries = [...histories];
-
-  // to organize data into rows to speed up hit-testing
-  const rows = [];
-
-  for (let row = 0; row < dates.length; row++) {
-    rows.push([]);
-  }
+export class GdprChart extends HTMLElement {
+  constructor() {
+    super();
+
+    this.attachShadow({mode: 'open'});
+
+    /** @type {Map<string, number[]>} */
+    this.histories = new Map();
+
+    /** @type {Date[]} */
+    this.dates = [];
+
+    /** @type {[string[], ...(Array<Array<number | null>>)]} */
+    this.rows = [[]];
+
+    this.hoverState = {seriesIndex: null, x: null, y: null};
 
-  let index = 0;
+    this.scrollContainer = document.createElement('div');
+    this.scrollContainer.classList.add('chart-scroll-container');
+    this.scrollContainer.addEventListener('scroll', this.onScroll.bind(this));
 
-  for (const [, history] of historyEntries) {
-    // each track is given one of 24 colours, which are spaced 15 degrees apart
-    // in hue
-    const hue = index * 15 % 360;
-    const color = `hsl(${hue},50%,33%)`;
-    const series = createSeries(history, color);
+    this.tooltip = this.createTooltip();
+
+    const stylesheet = document.createElement('link');
+    stylesheet.rel = 'stylesheet';
+    stylesheet.href = '/css/spotify-gdpr-chart.css';
+
+    this.shadowRoot.append(stylesheet);
+    this.shadowRoot.append(this.scrollContainer);
+    this.shadowRoot.append(this.tooltip);
+  }
+
+  /**
+   * Loads this GDPR chart with the given data.
+   *
+   * @param {Map<string, number[]>} histories The ranking history for each
+   * track.
+   * @param {Date[]} dates The date of each history entry.
+   * @public
+   */
+  load(histories, dates) {
+    this.histories = histories;
+    this.dates = dates;
+
+    // history + date info organized into rows for faster hit-testing
+    // first row is a header row containing keys
+    this.rows = [[...histories.keys()]];
 
     for (let row = 0; row < dates.length; row++) {
-      rows[row].push(history[row] || null);
+      this.rows.push([]);
     }
 
-    seriesContainer.append(series);
-    seriesElements.push(series);
-    index++;
-  }
+    for (const history of histories.values()) {
+      for (let row = 0; row < dates.length; row++) {
+        this.rows[row + 1].push(history[row] || null);
```

Comment:
At first I was wondering why you left the first row empty in `this.rows`, but now I see that you filled it on line 53.